0029-stream/0029-stream.md

```diff
@@ -170,19 +172,28 @@ Incoming packets can either be associated with an existing connection, or, for s
 
-Implementations SHOULD wait for a valid response (encrypted with the same shared secret) from the new address to validate the new path. STREAM uses the authenticated request/response packets in lieu of [QUIC's explicit Path Validation](https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#rfc.section.6.7). Implementations SHOULD refrain from sending large numbers of packets or large amounts of data to a new ILP address before validating the path to avoid being tricked into participating in a Denial of Service (DoS) attack on a third-party endpoint.
+#### 4.3.2. Connection Migration (Receivers Only)
 
-#### 4.3.1. Connection Asset Details
+If an endpoint supports receiving, then the endpoint MAY change its ILP Address at any point during a connection by sending a `ConnectionNewAddress` frame. To ensure the new address is received and acknowledged, implementations MAY choose to send these frames only in ILP Prepare packets, although certain connections may not support this (e.g., a receiver emitting this frame to a non-receiving sender will only be able to propagate this frame in a fulfill or reject packet).
 
-Each endpoint MAY expose their asset details by sending a `ConnectionAssetDetails` frame. This frame is optional because some use-cases do not require it.
+Senders encountering this frame SHOULD wait for a separate, valid request/response (encrypted with the same shared secret) from the new address to validate the new path. STREAM relies upon this authenticated request/response packet flow in lieu of [QUIC's explicit Path Validation](https://quicwg.github.io/base-drafts/draft-ietf-quic-transport.html#rfc.section.6.7), so implementations SHOULD refrain from sending large numbers of packets or large amounts of data to a new ILP address before validating the path. For example, this might help avoid being tricked into participating in a Denial of Service (DoS) attack on a third-party endpoint.
 
-Asset details exposed by this frame MUST not change during the lifetime of a Connection.
+#### 4.3.3. Connection Asset Details
+Either endpoint MAY expose its asset details by sending a `ConnectionAssetDetails` frame in a stream packet.
+
+Asset details, whether exposed by this frame or obtained by a higher-layer protocol, MUST not change during the lifetime of a Connection. Therefore, if a receiver receives a `ConnectionAssetDetails` frame that contradicts existing asset details, then the receiver SHOULD close the connection because it would be ambiguous which asset details are authoritative.
+
+This frame is OPTIONAL because the frame is generally only useful for senders who wish to verify the amount received on a path. For example, a client or server that only functions as a sender, but not a receiver, does not need to emit this frame if the endpoint on the other side of the connection does not need the information. Refer to Section 3.4 (Exchange Rates) for more information.
```

Comment:
Hmmm...I could be persuaded here. One thing about this RFC is that it leaves so much "to the reader" that it took me a very long time to realize why `ConnectionAssetDetails` are optional.

Put another way, in some cases the RFC doesn't do a good job of explaining why things are optional, which makes it difficult to know when something should or shouldn't be done -- especially as an implementor.

But maybe there's an argument there that a good RFC is just minimal language, and its purpose isn't to make it clear one way or another? If that's the intent of our RFCs, then I would be more apt to remove this type of added context paragraph.

Reply:
> Hmmm...I could be persuaded here. One thing about this RFC is that it leaves so much "to the reader" that it took me a very long time to realize why ConnectionAssetDetails are optional.

> especially as an implementor.

Agree that this RFC leaves _a lot_ to the implementer; it's much more of a specification than an implementer's guide. But I think there's so much nuance missing throughout this, maybe we'd need a separate document or notes for something like that?

I think a good analog is, could you easily implement a connector after reading the ILPv4 RFC? (Or, should you be able to?) Definitely not, but that's okay: the specification is still useful and relevant even if it's not clear how to apply it in a particular case.

What about something like this?

> Senders need the recipient to share their asset details in order to [enforce minimum exchange rates](link_to_3.4).

I'm not sure we should use more restrictive language than that since it's really never bad to send `ConnectionAssetDetails` -- even if you're only a sender (I worry more about implementations _not_ sending it when they should).

Reply:
Fixed via b488765bc2742ad40cebd34b56e2b0e254e76254